accord-core/src/main/java/accord/utils/async/AsyncResult.java

```diff
@@ -0,0 +1,96 @@
+package accord.utils.async;
+
+import java.util.concurrent.Executor;
+import java.util.function.BiConsumer;
+
+/**
+ * Handle for async computations that supports multiple listeners and registering
+ * listeners after the computation has started
+ */
+public interface AsyncResult<V>
```

Comment:
We spoke in slack but I feel that we should add `map` and `flatMap` here as well, so you are not forced to use the chain for single hop cases (which are 100% of the cases in Accord).

IMO we should replace `.toChain().map(...).complete()` with `.map()` and replace `.flatMap(...).flatMap(...).map(...)` with `.toChain().flatMap(...).flatMap(...).map(...).complete()`; AKA when you know you are a "chain" of calls, you call `toChain` to lower the cost.

I made this change in a feedback branch I can post, but the map/flatMap methods are

```
// feedback branch made sure null exeuctor works
default <T> AsyncResult<T> map(Function<? super V, ? extends T> mapper, Executor executor)
    {
        AsyncResult.Settable<T> settable = AsyncResults.settable();
        listen((success, failure) -> {
            if (failure != null)
            {
                settable.setFailure(failure);
                return;
            }
            try
            {
                T result = mapper.apply(success);
                settable.setSuccess(result);
            }
            catch (Throwable t)
            {
                settable.setFailure(t);
            }
        }, executor);
        return settable;
    }

    default <T> AsyncResult<T> map(Function<? super V, ? extends T> mapper)
    {
        return map(mapper, null);
    }

    default <T> AsyncResult<T> flatMap(Function<? super V, ? extends AsyncResult<T>> mapper, Executor executor)
    {
        AsyncResult.Settable<T> settable = AsyncResults.settable();
        listen((success, failure) -> {
           if (failure != null)
           {
               settable.setFailure(failure);
               return;
           }
           try
           {
               AsyncResult<T> next = mapper.apply(success);
               next.listen((s2, f2) -> {
                   if (f2 != null)
                   {
                       settable.tryFailure(f2);
                       return;
                   }
                   settable.trySuccess(s2);
               });
           }
           catch (Throwable t)
           {
               settable.setFailure(t);
           }
        }, executor);
        return settable;
    }

    default <T> AsyncResult<T> flatMap(Function<? super V, ? extends AsyncResult<T>> mapper)
    {
        return flatMap(mapper, null);
    }
```

I am totally cool with the trampoline (`AsyncChain`) but since the majority of cases are single hop we add a slightly higher cost than the raw `map` and `flatMap` methods (you can't avoid listening CAS, and the output Result has the same cost, so only difference is the cost to run the chain) I don't think we should default to use chain until we have at least 2 hops

Reply:
I asked to remove it, for two reasons: 1) it’s a lot of code duplication; 2) it encourages precisely what we’re trying to avoid. There’s no easy way to impose a “no more than one hop” restriction.

we could introduce some syntactic sugar for constructing a chain perhaps, that is then applied here, but I do not think these use cases warrant the escape route of dedicated map methods for the AsyncResult.

Reply:
> There’s no easy way to impose a “no more than one hop” restriction.

Based off usage, we don't have more than 1 hop... so are we trying to optimize for a case that we do not have?  I am looking at both accord-core and the C* integration branch, and every call to `toChain` yields a single hop.  We even have fun patterns such as this

```
private AsyncChain<Void> applyChain(SafeCommandStore safeStore, boolean canReschedule)
...
return applyWithCorrectScope(safeStore.commandStore()).toChain();

// usage
private AsyncResult<Void> applyWithCorrectScope(CommandStore unsafeStore)
...
AsyncChain<Void> chain = super.applyChain(safeStore);
result = AsyncResults.forChain(chain);
```

So, because of this, we go from `AsyncResult` to `AsyncChain` back to `AsyncResult`, costing more than just passing around the single `AsyncResult` (and require more cas/volatile overhead); this pattern happens several times...

Given we don't have multiple hops this argument doesn't hold for me; what are we really trying to solve here?

> it’s a lot of code duplication;

Is it?  Looking at my feedback branch and `AsyncChains` I really don't see duplicate code... they have duplicate signatures but not bodies.

Reply:
Wdym more cas/volatile overhead? There should be no more overhead in this regard. If there is, it’s a mistake and we should fix it.

Slightly different code performing the exact same work is code duplication?

Not having multiple hops today says nothing about the future? If you want to do chaining of AsyncResult, it would probably be better to have Future implement AshncResult and convert to Future, since at this point you are not saving anything with the separate concept.

Reply:
Yep, there should only be two CAS, same as we would need to directly transform. One to register the callback on the old AsyncResult, and one to set the new AsyncResult outcome. We have two intermediate objects: the Head (which is hopefully stack allocated), the Map node, and then the AsyncResult itself. So, is it the Map node you want to remove by merging with AsyncResult in the other case?

Reply:
If there's anything I would change, it would be the `beginAsResult` method to avoid the extra object invoking `settingCallback()` and instead have a variant of `AsyncResult.Settable` that is itself a callback. I would also consider renaming `toChain` as I think this gives the impression of more expense than there really is. HotSpot _should_ stack allocate the `Head` object, so we literally devolve to the same behaviour.

Reply:
> Wdym more cas/volatile overhead? There should be no more overhead in this regard. If there is, it’s a mistake and we should fix it.

We have a AsyncResult, we create a AsyncChain which listens to the result, we then produce a new AsyncResult which is a copy of the first one.  If we used the AsyncResult directly we would avoid the extra overhead (listen + copying result); see extra last comment going over this more clearly.

> Slightly different code performing the exact same work is code duplication?

In this case its the same semantic work, but that's a similar argument to saying `Iterator` shouldn't have a `forEach` because `Stream` has a `forEach`, or that `Routables.foldl` and `Topology.foldl` should not both exist.  Why is it a problem for `map` and `flatMap` to only exist in one type where as `forEach` and `fold` are allowed across multiple types?

> and not having multiple hops today says nothing about the future?

It does about the concern. If the concern is that we will have multiple hops, `AsyncChains` offers a way to lower this when we get this issue, but this has nothing to do with the majority of cases which only have a single hop.  Why choose to impact the majority of cases because one day we may have a case that could choose to use a different API call?

> If you want to do chaining of AsyncResult, it would probably be better to have Future implement AshncResult and convert to Future, since at this point you are not saving anything with the separate concept.

From a user/contributor to this project, `AsyncResult` to me is just a `Future`.  We are able to remove the cassandra dependency by taking ownership of our own future, which has its own benefit.  If we rename `AsyncResult` to `Future`, I am cool with this...

> Yep, there should only be two CAS, same as we would need to directly transform. One to register the callback on the old AsyncResult, and one to set the new AsyncResult outcome.

In the case of `map` they are equal, never attempted to argue that there are more CAS in this regard.  My comment about extra CAS was the new pattern of

```
private AsyncChain<Void> applyChain(SafeCommandStore safeStore, boolean canReschedule)
...
return applyWithCorrectScope(safeStore.commandStore()).toChain();

// usage
private AsyncResult<Void> applyWithCorrectScope(CommandStore unsafeStore)
...
AsyncChain<Void> chain = super.applyChain(safeStore);
result = AsyncResults.forChain(chain);
```

This pattern exists a few times in the Cassandra PR and are all the same:
1) We have a async operation, so have a `AsyncResult`
2) API forces `AsyncChain`, so "listen" to the result and return the chain
3) We want the result, so convert to `AsyncResult`

This pattern is where my comment comes from, this pattern "could" return the original result, but because we expose Chains we now have 2 extra CAS (1 to listen, 1 to store the result).

Reply:
From an efficiency perspective, I disagree this is material, as the bulk of cost can be resolved with an override to `beginAsResult`.

However, from an ergonomics perspective I agree this is suboptimal. I have pushed branches* that make `AsyncResult` and `AsyncChain` so that it can be used directly wherever an `AsyncChain` would be, and we avoid any conversion. The only semantic difference is that an `AsyncResult` may have an unlimited number of distinct `AsyncChain` started from them, so they create a kind of synchronisation point in the chain.

*
https://github.com/belliottsmith/cassandra-accord/tree/async-chains-suggest
https://github.com/belliottsmith/cassandra/tree/async-chains-suggest

Reply:
Benedict and I spoke in slack, and the summary is:

1) to simplify the usage, make `AsyncResult` implement `AsyncChain`, this has 2 benefits: no more `toChain`, and `AsyncChains.forChain` can avoid the copy of the result and just return the original result (avoiding 2 extra CAS; this case exists in C*)
2) AsyncChains.Head does not block multiple calls to `begin` which can allow multiple execution in Cassandra that were not possible before; we should block this.  This is localized to Head as `chain.map(...).beginAsResult()` would fail if begin was called multiple times